### PR TITLE
Feature: HMC5883L i2c compass

### PR DESCRIPTION
**Pull request by Joaomlg**
## Changes

1. **I2C1** configured for HMC5883L communication;
2. **HMC5883L Driver** layer implemented (setup configurations and read data);
3. **Compass Service** layer implemented (calibration and normalized degress angle calculation);
4. **USART2** configured for `printf` output stream (115200, 8n1);

## HMC5883L Driver usage example

```c
// [...]

HMC5883L_HandlerTypeDef compass_handler;
HMC5883L_DataTypeDef compass_data;

// [...]

// HMC5883 init
compass_handler.hi2c = &hi2c1;
compass_handler.samples_avg = HMC5883L_8_SAMPLES_AVG;
compass_handler.output_rate = HMC5883L_OUTPUT_RATE_75_HZ;
compass_handler.measurement_mode = HMC5883L_MEASUREMENT_MODE_NORMAL;
compass_handler.gain = HMC5883L_GAIN_4_7_GA;
compass_handler.i2c_speed = HMC5883L_I2C_NORMAL_SPEED;
compass_handler.operation_mode = HMC5883L_CONTINUOUS_MEASUREMENT_MODE;

HMC5883LDriver_Init(&compass_handler, 100);

// [...]

// while(1)
HMC5883LDriver_Read(&compass_handler, &compass_data, 100);
float angle = atan2f(compass_data.x_axis, compass_data.y_axis) * 180.00 / M_PI;

// [...]
```

### CODE REVIEW

File: Core/Src/Drivers/hmc5883l_driver.c [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+	if ((buffer[0] & 0x01) != 1) {
+		return;
+	}
+
+	HAL_I2C_Mem_Read(handler->hi2c, HMC5883l_ADDRESS, HMC5883l_DATAX_MSB_REG, 1, buffer, 6, timeout);
+
+	data->x_axis = (buffer[0] << 8) | buffer[1];
+	data->z_axis = (buffer[2] << 8) | buffer[3];
+	data->y_axis = (buffer[0] << 4) | buffer[5];
+}
```
Comment (MarianaLeite) [inline Core/Src/Drivers/hmc5883l_driver.c:50]: Let's make more clear when an error occurs. Here's a suggestion:

```suggestion
bool HMC5883LDriver_Read(HMC5883L_HandlerTypeDef *handler, HMC5883L_DataTypeDef *data, uint32_t timeout)
{
	uint8_t buffer[6];

	HAL_I2C_Mem_Read(handler->hi2c, HMC5883l_ADDRESS, HMC5883l_STATUS_REG, 1, buffer, 1, timeout);

	if ((buffer[0] & 0x01) != 1) {
		return false;
	}

	HAL_I2C_Mem_Read(handler->hi2c, HMC5883l_ADDRESS, HMC5883l_DATAX_MSB_REG, 1, buffer, 6, timeout);

	data->x_axis = (buffer[0] << 8) | buffer[1];
	data->z_axis = (buffer[2] << 8) | buffer[3];
	data->y_axis = (buffer[0] << 4) | buffer[5];
	
	return true;
}
```

To use `bool` you have to include `stdbool.h` header. Or if you prefer, we can create an unified return enumeration.

File: Core/Src/Services/compass_service.c [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+	calibration_params->x_offset = (calibration_params->xmax + calibration_params->xmin) / 2;
+	calibration_params->y_offset = (calibration_params->ymax + calibration_params->ymin) / 2;
+
+	x_var = calibration_params->xmax - calibration_params->xmin;
+	y_var = calibration_params->ymax - calibration_params->ymin;
+
+	if (x_var > y_var)
+	{
+		calibration_params->x_scale = 1.0;
+		calibration_params->y_scale = (float) x_var / y_var;
```
Comment (MarianaLeite) [inline Core/Src/Services/compass_service.c:47]: Please, ensure that there is no division by 0.

File: Core/Src/Services/compass_service.c [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+	}
+	else
+	{
+		calibration_params->y_scale = 1.0;
+		calibration_params->x_scale = (float) y_var / x_var;
```
Comment (MarianaLeite) [inline Core/Src/Services/compass_service.c:52]: Please, ensure that there is no division by 0.

File: Core/Src/main.c [hunk 4]
```diff
@@ -298,6 +412,30 @@ void L298N_Init(void)
...
+void HMC5883L_Calibrate(void)
+{
+	CompassService_InitCalibrationParams(&compass_calibration_params);
+	for (int i=0; i<1000; i++)
+	{
+		HMC5883LDriver_Read(&compass_handler, &compass_data, 100);
+		CompassService_Calibrate(compass_data.x_axis, compass_data.y_axis, &compass_calibration_params);
+		HAL_Delay(10);
+	}
+}
```
Comment (MarianaLeite) [inline Core/Src/main.c:438]: I think it would be better to move this to the CompassService.

File: Core/Src/main.c [hunk 5]
```diff
@@ -90,16 +119,34 @@ int main(void)
...
   /* Infinite loop */
   /* USER CODE BEGIN WHILE */
   L298NDriver_SetSpeed(&bridge_handler, 0.5);
   L298NDriver_MoveForward(&bridge_handler);
+
+//  HMC5883L_Calibrate();
+  compass_calibration_params.x_offset = 24;
+  compass_calibration_params.y_offset = 14;
+  compass_calibration_params.x_scale = 1.0;
+  compass_calibration_params.y_scale = (float)138/135;
```
Comment (MarianaLeite) [inline Core/Src/main.c:138]: These are default values? Why not placing them inside the `CompassService_InitCalibrationParams` function?

Comment (Joaomlg) [reply Core/Src/main.c:137]: `CompassService_InitCalibrationParams` function setup the structure parameters to start a new calibration... These values ​​that are there were obtained when the current hardware was calibrated, and will change from different hardwares.

**Review by MarianaLeite — APPROVED** [review]

**Pull request merged**